backend/icloud/icloud.go

```diff
@@ -0,0 +1,1149 @@
+	}
+
+	item, found, err := f.findLeafItem(ctx, ID, leaf)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if !found {
+		return nil, fs.ErrorObjectNotFound
+	}
+
+	return item, nil
+}
+
+func (o *Object) setMetaData(item *api.DriveItem) (err error) {
+	if item.IsFolder() {
+		return fs.ErrorIsDir
+	}
+	o.size = item.Size
+	o.modTime = item.DateModified
+	o.createdTime = item.DateCreated
+	o.driveID = item.Drivewsid
+	o.docID = item.Docwsid
+	o.itemID = item.Itemid
+	o.etag = item.Etag
+	o.downloadURL = item.DownloadURL()
+	return nil
+}
+
+// ID returns the ID of the Object if known, or "" if not
+func (o *Object) ID() string {
+	return o.docID
+}
+
+// Fs implements fs.Object.
+func (o *Object) Fs() fs.Info {
+	return o.fs
+}
+
+// Hash implements fs.Object.
+func (o *Object) Hash(ctx context.Context, t hash.Type) (string, error) {
+	return "", hash.ErrUnsupported
+}
+
+// ModTime implements fs.Object.
+func (o *Object) ModTime(context.Context) time.Time {
+	return o.modTime
+}
+
+// Open implements fs.Object.
+func (o *Object) Open(ctx context.Context, options ...fs.OpenOption) (io.ReadCloser, error) {
+	fs.FixRangeOption(options, o.size)
+
+	// Drive does not support empty files, so we cheat
+	if o.size == 0 {
+		return io.NopCloser(bytes.NewBufferString("")), nil
+	}
+
+	var resp *http.Response
+	var err error
+
+	if err = o.fs.pacer.Call(func() (bool, error) {
+		var url string
+ 
+    //var doc *api.Document
+		//if o.docID == "" {
+			//doc, resp, err = o.fs.service.GetDocByItemID(ctx, o.itemID)
+    //}
+    
+    // Can not get the download url on a item to work, so do it the hard way.
+		url, _, err = o.fs.service.GetDownloadURLByDriveID(ctx, o.driveID)
+
+		resp, err = o.fs.service.DownloadFile(ctx, url, options)
+		return shouldRetry(ctx, resp, err)
+	}); err != nil {
+		return nil, err
+	}
+
+	return resp.Body, err
+}
+
+// Remote implements fs.Object.
+func (o *Object) Remote() string {
+	return o.remote
+}
+
+// Remove implements fs.Object.
+func (o *Object) Remove(ctx context.Context) error {
+	if o.itemID == "" {
+		return nil
+	}
+
+	var resp *http.Response
+	var err error
+	if err = o.fs.pacer.Call(func() (bool, error) {
+		_, resp, err = o.fs.service.MoveItemToTrashByItemID(ctx, o.itemID, o.etag, true)
+		return retryResultUnknown(ctx, resp, err)
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetModTime implements fs.Object.
+func (o *Object) SetModTime(ctx context.Context, t time.Time) error {
+	return fs.ErrorCantSetModTime
+}
+
+// Size implements fs.Object.
+func (o *Object) Size() int64 {
+	return o.size
+}
+
+// Storable implements fs.Object.
+func (o *Object) Storable() bool {
+	return true
+}
+
+// String implements fs.Object.
+func (o *Object) String() string {
+	if o == nil {
+		return "<nil>"
+	}
+	return o.remote
+}
+
+// Update implements fs.Object.
+// TODO: Implement restoring the old file when an errror occures during upload
+// TODO: Implement removing old file from trash when upload is succesfull
```

Comment:
Rclone policy is that we use the trash by default but supply a parameter called `hard_delete` for backends that can trash stuff and permanently delete stuff.

Very early versions of rclone hard delete by default, but we had a few data loss incidents so we changed the policy!

Reply:
k yeah trash is the default. Im
low on time tho so ill keep the implementation of hard delete for another day